src/main/java/com/google/sps/servlets/BooksServlet.java

```diff
@@ -0,0 +1,151 @@
+    }
+    String json = makeBooksJson();
+    response.getWriter().println(json);
+  }
+
+ /**
+  * Encodes the given term so that it can fit in a url.
+  * @return An encoded string
+  */
+  private String encodeTerm(String term) {
+    try {
+      return URLEncoder.encode(term, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
+  private String makeBooksJson() {
+    Gson gson = new Gson();
+    Map<String, Object> books = new HashMap<>();
+    books.put("books", booksMap);
+    String json = gson.toJson(books); 
+    return json;
+  }
+
+ /**
+  * Adds Book object to books from JSON results of Guardian query.
+  */
+  private void jsonToBooks(String jsonString, String keyTerm) {
+    JsonObject responseObject = new JsonParser().parse(jsonString).getAsJsonObject();
+    // Check to see if json has enough entries.
+    int numResults = responseObject.get("totalItems").getAsInt();
+    int resultsToShow = NUM_BOOKS_PER_KEYWORD;
+    if (numResults < resultsToShow) {
+      if (numResults > 0) {
+        resultsToShow = numResults;
+      } else {
+        System.err.println("Error: no results found for query");
+        return;
+      }
+    }
+    JsonArray booksJsonArray = responseObject.getAsJsonArray("items");
+    List<Book> books = new ArrayList<>();
+    for (int i = 0; i < resultsToShow; i++) {
+      JsonObject bookJson = booksJsonArray.get(i).getAsJsonObject().getAsJsonObject("volumeInfo"); 
+      String title = bookJson.get("title").getAsString();
+      String link = bookJson.get("infoLink").getAsString();
+      String image = bookJson.getAsJsonObject("imageLinks").get("thumbnail").getAsString();
+      String description = bookJson.get("description").getAsString();
+      String writer = formatWriters(bookJson.getAsJsonArray("authors"));
+      books.add(new Book.Builder(title, link).withImage(image).withDescription(description).withWriter(writer).build());
+    }
+    booksMap.put(keyTerm, books);
+  }
+
+ /**
+  * Reads in JSON from an open URL.
+  * @return JSON in String format.
+  */
+  private String getJson(URL url) throws IOException {
+    Scanner jsonReader = new Scanner(url.openStream());
+    String data = "";
+    while(jsonReader.hasNext()) {
+        data += jsonReader.nextLine();
+    }
+    jsonReader.close();
+    return data;
+  }
+
+ /**
+  * Formats writers into string from JSON array.
+  * @return writers in String fomat.
+  */
+  private String formatWriters(JsonArray writersArray) {
+    String returnString = "";
+    int length = writersArray.size();
+    for (int i = 0; i < length - 1; i++){
+      returnString += writersArray.get(i).getAsString() + ", ";
+    }
+    returnString += writersArray.get(length - 1).getAsString();
+    return returnString;
+  }
+
+ /**
+  * Adds a Book object in books for a key term.
+  */
+  private void addBooksForTerm(String keyTerm, String encodedKeyTerm) {
```

Comment:
[Optional] Instead of passing both arguments, you could pass only keyTerm and encode it within the method.